### ISSUE TITLE
Conversion error when loading a model with v0.13+ with BSON

### ISSUE THREAD

**Issue by sloede** [FluxML/Flux.jl#1984]
We are using Flux.jl to identify discontinuous solutions in flow simulations in [Trixi.jl](https://www.github.com/trixi-framework/Trixi.jl). When trying to load a Flux.jl model that was previously generated and saved using BSON with Flux.jl v0.13 (trixi-framework/Trixi.jl#1155), we get the following error:
```julia
ERROR: LoadError: MethodError: Cannot `convert` an object of type Float64 to an object of type Vector{Any}
Closest candidates are:
  convert(::Type{T}, ::LinearAlgebra.Factorization) where T<:AbstractArray at /mnt/hd1/opt/julia/1.7.2/share/julia/stdlib/v1.7/LinearAlgebra/src/factorization.jl:58
  convert(::Type{Array{T, N}}, ::FillArrays.Zeros{V, N}) where {T, V, N} at ~/.julia/packages/FillArrays/5Arin/src/FillArrays.jl:440
  convert(::Type{Array{T, N}}, ::FillArrays.Ones{V, N}) where {T, V, N} at ~/.julia/packages/FillArrays/5Arin/src/FillArrays.jl:440
  ...
Stacktrace:
  [1] newstruct!(::IdDict{Any, Any}, ::Float64, ::Function, ::Nothing)
    @ BSON ~/.julia/packages/BSON/rOaki/src/extensions.jl:107
  [2] newstruct_raw(cache::IdDict{Any, Any}, T::Type, d::Dict{Symbol, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/extensions.jl:154
[...]
wololo
```
<details>
<summary>Full error message</summary>

```julia
ERROR: LoadError: MethodError: Cannot `convert` an object of type Float64 to an object of type Vector{Any}
Closest candidates are:
  convert(::Type{T}, ::LinearAlgebra.Factorization) where T<:AbstractArray at /mnt/hd1/opt/julia/1.7.2/share/julia/stdlib/v1.7/LinearAlgebra/src/factorization.jl:58
  convert(::Type{Array{T, N}}, ::FillArrays.Zeros{V, N}) where {T, V, N} at ~/.julia/packages/FillArrays/5Arin/src/FillArrays.jl:440
  convert(::Type{Array{T, N}}, ::FillArrays.Ones{V, N}) where {T, V, N} at ~/.julia/packages/FillArrays/5Arin/src/FillArrays.jl:440
  ...
Stacktrace:
  [1] newstruct!(::IdDict{Any, Any}, ::Float64, ::Function, ::Nothing)
    @ BSON ~/.julia/packages/BSON/rOaki/src/extensions.jl:107
  [2] newstruct_raw(cache::IdDict{Any, Any}, T::Type, d::Dict{Symbol, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/extensions.jl:154
  [3] (::BSON.var"#49#50")(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/extensions.jl:169
  [4] raise_recursive(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:92
  [5] (::BSON.var"#23#24"{IdDict{Any, Any}, Module})(x::Dict{Symbol, Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:98
  [6] applychildren!(f::BSON.var"#23#24"{IdDict{Any, Any}, Module}, x::Vector{Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/BSON.jl:26
  [7] raise_recursive
    @ ~/.julia/packages/BSON/rOaki/src/read.jl:98 [inlined]
  [8] (::BSON.var"#18#21"{IdDict{Any, Any}, Module})(x::Vector{Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:82
  [9] applychildren!(f::BSON.var"#18#21"{IdDict{Any, Any}, Module}, x::Dict{Symbol, Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/BSON.jl:19
 [10] _raise_recursive(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:82
 [11] raise_recursive(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:93
 [12] (::BSON.var"#23#24"{IdDict{Any, Any}, Module})(x::Dict{Symbol, Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:98
 [13] applychildren!(f::BSON.var"#23#24"{IdDict{Any, Any}, Module}, x::Vector{Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/BSON.jl:26
 [14] raise_recursive
    @ ~/.julia/packages/BSON/rOaki/src/read.jl:98 [inlined]
 [15] (::BSON.var"#18#21"{IdDict{Any, Any}, Module})(x::Vector{Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:82
 [16] applychildren!(f::BSON.var"#18#21"{IdDict{Any, Any}, Module}, x::Dict{Symbol, Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/BSON.jl:19
 [17] _raise_recursive(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:82
 [18] (::BSON.var"#49#50")(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/extensions.jl:168
 [19] raise_recursive(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:92
 [20] (::BSON.var"#19#22"{IdDict{Any, Any}, Module})(x::Dict{Symbol, Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:86
 [21] applychildren!(f::BSON.var"#19#22"{IdDict{Any, Any}, Module}, x::Dict{Symbol, Any})
    @ BSON ~/.julia/packages/BSON/rOaki/src/BSON.jl:19
 [22] _raise_recursive(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:86
 [23] raise_recursive(d::Dict{Symbol, Any}, cache::IdDict{Any, Any}, init::Module)
    @ BSON ~/.julia/packages/BSON/rOaki/src/read.jl:93
 [24] raise_recursive
    @ ~/.julia/packages/BSON/rOaki/src/read.jl:103 [inlined]
 [25] load (repeats 2 times)
    @ ~/.julia/packages/BSON/rOaki/src/read.jl:108 [inlined]
 [26] top-level scope
    @ /mnt/ssd/home/mschlott/hackathon/bump-compat/tmp2/reproduce.jl:8
in expression starting at /mnt/ssd/home/mschlott/hackathon/bump-compat/tmp2/reproduce.jl:8
```
</details>

We can reproduce this issue with the following MWE:

`reproduce.jl`:
```julia
# reproduce.jl
using Downloads: Downloads
using BSON: BSON
using Flux
using NNlib

network = joinpath(@__DIR__, "modelnnrhs-0.973-0.001.bson")
Downloads.download("https://gist.github.com/JuliaOd/97728c2c15d6a7255ced6e46e3a605b6/raw/modelnnrhs-0.973-0.001.bson", network)

bload = BSON.load(network)
println(bload)
```

With the last Flux.jl version in the v0.12 family
```shell
julia --project=. -e 'using Pkg; Pkg.add(["BSON", "NNlib"]); Pkg.add(name="Flux", version="0.12.10")'
julia --project reproduce.jl
```
we get the expected output
```julia
Dict{Symbol, Any}(:model2d => Chain(Dense(15, 20, leakyrelu), Dropout(0.01), Dense(20, 20, leakyrelu), Dropout(0.01), Dense(20, 20, leakyrelu), Dropout(0.01), Dense(20, 20, leakyrelu), Dropout(0.01), Dense(20, 20, leakyrelu), Dense(20, 2), softmax))
```

However, with Flux.jl v0.13.0
```shell
julia --project=. -e 'using Pkg; Pkg.add(["BSON", "NNlib"]); Pkg.add(name="Flux", version="0.13.0")'
julia --project reproduce.jl
```
we obtain the described error above.

Since I am completely at a loss as to what is broken and how to debug this, it would be great if someone could give me some pointers on how to proceed.

cc @ranocha

**Comment by ToucheSir** [comment]
This is a otherwise non-breaking change in the [type parameters](https://github.com/FluxML/Flux.jl/blob/v0.13.3/src/layers/normalise.jl#L69) of `Flux.Dropout` which is causing issues in BSON.
It's unclear to me at this point whether the BSON problem is a simple bugfix or a more fundamental issue.

In the meantime, you can employ the following workaround:

1. Load your model in Flux 0.12.x.
2. Replace all `Dropout` layers with a custom struct containing the same fields. The easisest way to do this is via `fmap` with a custom `exclude` function, so happy to provide code if you need help there.
3. Save the modified model and read it back in on Flux 0.13
3. Replace all the custom structs with `Dropout`.

**Comment by sloede** [comment]
Thanks for the quick reply. I tried to put something together that *seems* to work - it would be great if you could have a quick look and point out obvious flaws/errors in my logic.

Custom struct definition including conversion ctors in `tmpdropout.jl`:
```julia
using Flux

# Create custom Dropout type for conversion
mutable struct TmpDropout{F,D}
  p::F
  dims::D
  active::Union{Bool, Nothing}
end
TmpDropout(d::Flux.Dropout) = TmpDropout(d.p, d.dims, d.active)
Flux.Dropout(d::TmpDropout) = Dropout(d.p, d.dims, d.active)
```
Then, I load the Flux.jl model from the old BSON in Flux v0.12, convert it using something like
```julia
fmap(TmpDropout, model; exclude=x->x isa Dropout)
```
store it in a temporary BSON, then restart Julia, add Flux v0.13, load the model from temporary BSON, convert it using something like
```julia
fmap(Dropout, model; exclude=x->x isa TmpDropout)
```
and store it in the final BSON. Is this what you had in mind?

For reference, I have included the actual conversion files I used below:
<details>
<summary>Full example</summary>

Conversion of existing BSON to temporary BSON with `TmpDropout` in `convert-v0.12-to-TmpDropout.jl`, using Flux **v0.12**:
```julia
using Downloads: Downloads
using BSON: BSON
using Flux

include("tmpdropout.jl")

# Get the original BSON file
network = joinpath(@__DIR__, "modelnnrhs-0.973-0.001.bson")
Downloads.download("https://gist.github.com/JuliaOd/97728c2c15d6a7255ced6e46e3a605b6/raw/modelnnrhs-0.973-0.001.bson", network)

# Load the model from BSON
model2d = BSON.load(network, @__MODULE__)[:model2d]

# Convert model to use TmpDropout instead of Flux.Dropout
model2d_tmpdropout = fmap(TmpDropout, model2d; exclude=x->x isa Dropout)

# Store model to BSON
BSON.bson("modelnnrhs-0.973-0.001-TmpDropout.bson", Dict{Symbol,Any}(:model2d=>model2d_tmpdropout))
```

Conversion of temporary BSON to new BSON with `Flux.Dropout` in `convert-TmpDropout-to-v0.13.jl`, using Flux **v0.13**:
```julia
using BSON: BSON
using Flux

include("tmpdropout.jl")

# Load the model that uses TmpDropout from BSON
network = joinpath(@__DIR__, "modelnnrhs-0.973-0.001-TmpDropout.bson")
model2d_tmpdropout = BSON.load(network, @__MODULE__)[:model2d]

# Convert model to use new Flux.Dropout instead of TmpDropout
model2d = fmap(Dropout, model2d_tmpdropout; exclude=x->x isa TmpDropout)

# Store model to BSON
BSON.bson("modelnnrhs-0.973-0.001-v0.13.bson", Dict{Symbol,Any}(:model2d=>model2d))
```
</details>

**Comment by ToucheSir** [comment]
Yes it is. You'll want to check that the model is still functional after conversion because `fmap` does recurse pretty far down by default, but if it is then everything should be good.

**Comment by sloede** [comment]
Thanks again for all your help! The conversion seems to have worked, we were able to merge the relevant PR without having to change any test results. I commend you on your efforts in #1988, I think this would be very valuable. You might also want to consider to add something to the docs (e.g., [here](https://fluxml.ai/Flux.jl/stable/saving/#Saving-and-Loading-Models)) that makes users aware that while BSON itself is stable enough, internal changes in Flux might break compatibility between stored models but that there is a workaround until a proper solution is found.

Closing this now.
